File: RailgunVB/Core/Preconditions/UserPerms.vb
Imports Discord
Imports Discord.Commands
Imports Microsoft.Extensions.DependencyInjection
Imports RailgunVB.Core.Configuration

Namespace Core.Preconditions
    
    <AttributeUsage(AttributeTargets.Class Or AttributeTargets.Method, AllowMultiple := True)>
    Public NotInheritable Class UserPerms
        Inherits PreconditionAttribute
    
        Private ReadOnly _guildPermission As GuildPermission? = Nothing
        Private ReadOnly _channelPermission As ChannelPermission? = Nothing
        
        Public Sub New(guildPermission As GuildPermission)
             _guildPermission = guildPermission
        End Sub
        
        Public Sub New(channelPermission As ChannelPermission)
            _channelPermission = channelPermission
        End Sub

        Public Overrides Async Function CheckPermissionsAsync(context As ICommandContext, command As CommandInfo, services As IServiceProvider) As Task(Of PreconditionResult)
            Dim config As MasterConfig = services.GetService(Of MasterConfig)
            Dim user As IGuildUser = Await context.Guild.GetUserAsync(context.User.Id)
            
            If _guildPermission.HasValue
                If Not (user.GuildPermissions.Has(_guildPermission) OrElse user.GuildPermissions.Administrator OrElse 
                   config.DiscordConfig.OtherAdmins.Contains(user.Id) OrElse 
                   user.Id = config.DiscordConfig.MasterAdminId)
                    Return Await Task.FromResult(PreconditionResult.FromError(
                        $"You do not have permission to use this command! {Format.Bold(
                            $"SERVER-PERM-MISSING : {_guildPermission.ToString()}")}"))
                End If
            ElseIf _channelPermission.HasValue
                Dim channelPerms As ChannelPermissions = user.GetPermissions(context.Channel)
                
                If Not (channelPerms.Has(_channelPermission.Value) OrElse user.GuildPermissions.Administrator OrElse 
                        config.DiscordConfig.OtherAdmins.Contains(user.Id) OrElse 
                        user.Id = config.DiscordConfig.MasterAdminId) 
                    Return Await Task.FromResult(PreconditionResult.FromError(
                        $"You do not have permission to use this command! {Format.Bold(
                            $"CHANNEL-PERM-MISSING : {_channelPermission.ToString()}")}"))
                End If
            End If
            
            Return Await Task.FromResult(PreconditionResult.FromSuccess())
        End Function
        
    End Class
    
End NameSpace

File: Railgun/Core/Preconditions/UserPerms.cs
using System;
using System.Threading.Tasks;
using Discord;
using Discord.Commands;
using Microsoft.Extensions.DependencyInjection;
using Railgun.Core.Configuration;

namespace Railgun.Core.Preconditions
{
    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = true)]
    public class UserPerms : PreconditionAttribute
    {
        private readonly GuildPermission? _guildPermission = null;
        private readonly ChannelPermission? _channelPermission = null;

        public UserPerms(GuildPermission permission) => _guildPermission = permission;

        public UserPerms(ChannelPermission permission) => _channelPermission = permission;

        public override async Task<PreconditionResult> CheckPermissionsAsync(ICommandContext context, CommandInfo command, IServiceProvider services) {
            var config = services.GetService<MasterConfig>();
            var user = await context.Guild.GetUserAsync(context.User.Id);

            if (_guildPermission.HasValue) {
                if (!(user.GuildPermissions.Has(_guildPermission.Value) || user.GuildPermissions.Administrator || config.DiscordConfig.OtherAdmins.Contains(user.Id) || user.Id == config.DiscordConfig.MasterAdminId))
                    return await Task.FromResult(PreconditionResult.FromError($"You do not have permission to use this command! {Format.Bold($"SERVER-PERM-MISSING : {_guildPermission.ToString()}")}"));
            } else if (_channelPermission.HasValue) {
                var channelPerms = user.GetPermissions((IGuildChannel)context.Channel);

                if (!(channelPerms.Has(_channelPermission.Value) || user.GuildPermissions.Administrator || config.DiscordConfig.OtherAdmins.Contains(user.Id) || user.Id == config.DiscordConfig.MasterAdminId))
                    return await Task.FromResult(PreconditionResult.FromError($"You do not have permission to use this command! {Format.Bold($"CHANNEL-PERM-MISSING : {_channelPermission.ToString()}")}"));
            }

            return await Task.FromResult(PreconditionResult.FromSuccess());
        }
    }
}